libraries/AP_HAL_PX4/CAN.cpp

```diff
@@ -0,0 +1,1035 @@
+
+    return true; // This new frame will be added to a free TX mailbox in the next @ref send().
+}
+
+bool PX4CAN::isRxBufferEmpty() const
+{
+    CriticalSectionLocker lock;
+    return rx_queue_.available() == 0;
+}
+
+uint64_t PX4CAN::getErrorCount() const
+{
+    CriticalSectionLocker lock;
+    return error_cnt_;
+    //TODO: + rx_queue_.getOverflowCount();
+}
+
+unsigned PX4CAN::getRxQueueLength() const
+{
+    CriticalSectionLocker lock;
+    return rx_queue_.available();
+}
+
+bool PX4CAN::hadActivity()
+{
+    CriticalSectionLocker lock;
+    const bool ret = had_activity_;
+    had_activity_ = false;
+    return ret;
+}
+
+bool PX4CAN::begin(uint32_t bitrate)
+{
+    if (init(bitrate, OperatingMode::NormalMode) == 0) {
+        bitrate_ = bitrate;
+        initialized_ = true;
+    } else {
+        initialized_ = false;
+    }
+    return initialized_;
+}
+
+void PX4CAN::end()
+{
+}
+
+void PX4CAN::reset()
+{
+    if (initialized_ && bitrate_ != 0) {
+        init(bitrate_, OperatingMode::NormalMode);
+    }
+}
+
+bool PX4CAN::is_initialized()
+{
+    return initialized_;
+}
+
+int32_t PX4CAN::available()
+{
+    if (initialized_) {
+        return getRxQueueLength();
+    } else {
+        return -1;
+    }
+}
+
+int32_t PX4CAN::tx_pending()
+{
+    int32_t ret = -1;
+
+    if (initialized_) {
+        ret = 0;
+        CriticalSectionLocker lock;
+
+        for (int mbx = 0; mbx < NumTxMailboxes; mbx++) {
+            if (pending_tx_[mbx].pending) {
+                ret++;
+            }
+        }
+    }
+
+    return ret;
+}
+
+/*
+ * CanDriver
+ */
+uavcan::CanSelectMasks PX4CANDriver::makeSelectMasks(const uavcan::CanFrame* (&pending_tx)[uavcan::MaxCanIfaces]) const
+{
+    uavcan::CanSelectMasks msk;
+
+    if (can_number_ >= 1) {
+        // Iface 0
+        msk.read = if0_.isRxBufferEmpty() ? 0 : 1;
+
+        if (pending_tx[0] != UAVCAN_NULLPTR) {
+            msk.write = if0_.canAcceptNewTxFrame(*pending_tx[0]) ? 1 : 0;
+        }
+    }
+
+    if (can_number_ >= 2) {
+        if (!if1_.isRxBufferEmpty()) {
+            msk.read |= 1 << 1;
+        }
+
+        // Iface 1
+        if (pending_tx[1] != UAVCAN_NULLPTR) {
+            if (if1_.canAcceptNewTxFrame(*pending_tx[1])) {
+                msk.write |= 1 << 1;
+            }
+        }
+    }
+
+    return msk;
+}
+
+bool PX4CANDriver::hasReadableInterfaces() const
+{
+    if (can_number_ >= 2) {
+        return !if0_.isRxBufferEmpty() || !if1_.isRxBufferEmpty();
+    }
+
+    return !if0_.isRxBufferEmpty();
+}
+
+int16_t PX4CANDriver::select(uavcan::CanSelectMasks& inout_masks,
+                             const uavcan::CanFrame* (&pending_tx)[uavcan::MaxCanIfaces], const uavcan::MonotonicTime blocking_deadline)
+{
+    const uavcan::CanSelectMasks in_masks = inout_masks;
+    const uavcan::MonotonicTime time = clock::getMonotonic();
+
+    if (can_number_ >= 1) {
+        if0_.discardTimedOutTxMailboxes(time); // Check TX timeouts - this may release some TX slots
```

Comment:
Having two small critical sections is preferable over having one large critical section because the latter case provides for better worst case IRQ latency, because a pending IRQ can be handled between the two critical sections.